USB STM32: Improve usb clock code

kbidani opened the pull request:
This PR includes several improvements to the USB code for STM32 devices. The changes aim to enhance code maintainability and add necessary configurations for PHY clocks.

kbidani committed:
drivers: usb: udc_stm32: imporve usb code maintainability

Replace the direct call to LL_AHBx_GRP1_EnableClock
with the clock control API.

Signed-off-by: Khaoula Bidani <khaoula.bidani-ext@st.com>

kbidani committed:
dts: arm: st: add phy clocks to dtsi

add phy clocks to dtsi files.

Signed-off-by: Khaoula Bidani <khaoula.bidani-ext@st.com>

kbidani committed:
drivers: usb: device_stm32: imporve usb code maintainability

Replace the direct call to LL_AHBx_GRP1_EnableClock
with the clock control API.

Signed-off-by: Khaoula Bidani <khaoula.bidani-ext@st.com>

kbidani committed:
boards: adi: sdp_k1: add phy clocks to dtsi

add phy clocks to dtsi files.

Signed-off-by: Khaoula Bidani <khaoula.bidani-ext@st.com>

etienne-lms commented on dts/arm/st/h7/stm32h723.dtsi:
```diff
@@ -234,6 +234,7 @@
 	otghs_fs_phy: otghs_fs_phy {
 		compatible = "usb-nop-xceiv";
 		#phy-cells = <0>;
+		clocks = <&rcc STM32_CLOCK(AHB2, 31U)>;
```
Could you crosscheck that? It's seems to me USB1OTGHSULPIEN is located in RCC_AHB1xxR[bit 26] on stm32h723/h725.

mathieuchopstm replied:
```suggestion
		clocks = <&rcc STM32_CLOCK(AHB2, 31)>;
```
(+ change the value according to Etienne's remark if applicable)

etienne-lms commented on drivers/usb/udc/udc_stm32.c:
```diff
@@ -1164,7 +1160,7 @@ static int priv_clock_enable(void)
 
 #if USB_OTG_HS_EMB_PHY
-	LL_APB2_GRP1_EnableClock(LL_APB2_GRP1_PERIPH_OTGPHYC);
+	clock_control_on(clk, (clock_control_subsys_t)&phy_pclken[0]);
```
also change U5 case at line 1184?

etienne-lms commented on drivers/usb/device/usb_dc_stm32.c:
```diff
@@ -89,6 +89,12 @@ static const struct gpio_dt_spec ulpi_reset =
 
+#if USB_OTG_HS_ULPI_PHY || \
+DT_HAS_COMPAT_STATUS_OKAY(st_stm32u5_otghs_phy) || \
+USB_OTG_HS_EMB_PHYC
```
Indent

etienne-lms replied:
Actually here also `#if USB_OTG_HS_ULPI_PHY || USB_OTG_HS_EMB_PHYC` should be enough.

etienne-lms commented on drivers/usb/device/usb_dc_stm32.c:
```diff
@@ -211,9 +217,8 @@ void HAL_PCD_SOFCallback(PCD_HandleTypeDef *hpcd)
-#if DT_HAS_COMPAT_STATUS_OKAY(st_stm32u5_otghs_phy)
 
-static const struct stm32_pclken phy_pclken[] = STM32_DT_CLOCKS(DT_INST_PHANDLE(0, phys));
+#if DT_HAS_COMPAT_STATUS_OKAY(st_stm32u5_otghs_phy)
```
nitpicking: remove the extra empty line above.

etienne-lms commented on drivers/usb/device/usb_dc_stm32.c:
```diff
@@ -389,28 +394,24 @@ static int usb_dc_stm32_clock_enable(void)
-	LL_AHB2_GRP1_EnableClock(LL_AHB2_GRP1_PERIPH_USBPHY);
-	/* Both OTG HS and USBPHY sleep clock MUST be disabled here at the same time */
+	clock_control_on(clk, (clock_control_subsys_t)&phy_pclken[0]);
+/* Both OTG HS and USBPHY sleep clock MUST be disabled here at the same time */
```
indentation

etienne-lms commented on boards/arduino/portenta_h7/arduino_portenta_h7-common.dtsi:
```diff
@@ -22,9 +22,9 @@
+		clocks = <&rcc STM32_CLOCK(AHB1, 26U)>;
 		reset-gpios = < &gpioj 4 GPIO_ACTIVE_LOW >;
 		#phy-cells = <0>;
-		clocks = <&rcc STM32_CLOCK(AHB1, 26U)>;
```
Can discard this change (no change)

mathieuchopstm commented on boards/adi/sdp_k1/adi_sdp_k1.dts:
```diff
@@ -45,6 +45,7 @@
 	otghs_ulpi_phy: otghs_ulpis_phy {
 		compatible = "usb-ulpi-phy";
 		reset-gpios = <&gpiod 7 (GPIO_ACTIVE_LOW)>;
+		clocks = <&rcc STM32_CLOCK(AHB1, 30U)>;
```
Do not use `U` suffix for Device Tree constants. Ditto every file.
```suggestion
		clocks = <&rcc STM32_CLOCK(AHB1, 30)>;
```

mathieuchopstm commented on boards/arduino/nicla_vision/arduino_nicla_vision_stm32h747xx_m7.dts:
```diff
@@ -32,6 +32,7 @@
 
 	otghs_ulpi_phy: otghs_ulpis_phy {
 		compatible = "usb-ulpi-phy";
+		clocks = <&rcc STM32_CLOCK(AHB1, 26U)>;
```
```suggestion
		clocks = <&rcc STM32_CLOCK(AHB1, 26)>;
```

mathieuchopstm commented on boards/arduino/portenta_h7/arduino_portenta_h7-common.dtsi:
```diff
@@ -22,6 +22,7 @@
 
 	otghs_ulpi_phy: otghs_ulpis_phy {
 		compatible = "usb-ulpi-phy";
+		clocks = <&rcc STM32_CLOCK(AHB1, 26U)>;
```
```suggestion
		clocks = <&rcc STM32_CLOCK(AHB1, 26)>;
```

mathieuchopstm commented on boards/st/stm32f769i_disco/stm32f769i_disco.dts:
```diff
@@ -36,6 +36,7 @@
 	otghs_ulpi_phy: otghs_ulpis_phy {
 		compatible = "usb-ulpi-phy";
 		#phy-cells = <0>;
+		clocks = <&rcc STM32_CLOCK(AHB1, 30U)>;
```
```suggestion
		clocks = <&rcc STM32_CLOCK(AHB1, 30)>;
```

mathieuchopstm commented on boards/st/stm32h757i_eval/stm32h757i_eval_stm32h757xx_m7.dts:
```diff
@@ -73,6 +73,7 @@
 	otghs_ulpi_phy: otghs_ulpis_phy {
 		compatible = "usb-ulpi-phy";
 		#phy-cells = <0>;
+		clocks = <&rcc STM32_CLOCK(AHB1, 26U)>;
```
```suggestion
		clocks = <&rcc STM32_CLOCK(AHB1, 26)>;
```

mathieuchopstm commented on boards/st/stm32h747i_disco/stm32h747i_disco_stm32h747xx_m7.dts:
```diff
@@ -57,6 +57,7 @@
 	otghs_ulpi_phy: otghs_ulpis_phy {
 		compatible = "usb-ulpi-phy";
 		#phy-cells = <0>;
+		clocks = <&rcc STM32_CLOCK(AHB1, 26U)>;
```
```suggestion
		clocks = <&rcc STM32_CLOCK(AHB1, 26)>;
```

mathieuchopstm commented on drivers/usb/device/usb_dc_stm32.c:
```diff
@@ -389,28 +394,24 @@ static int usb_dc_stm32_clock_enable(void)
-	/* Both OTG HS and USBPHY sleep clock MUST be disabled here at the same time */
+	clock_control_on(clk, (clock_control_subsys_t)&phy_pclken[0]);
+/* Both OTG HS and USBPHY sleep clock MUST be disabled here at the same time */
 	LL_AHB2_GRP1_DisableClockStopSleep(LL_AHB2_GRP1_PERIPH_OTG_HS ||
 						LL_AHB2_GRP1_PERIPH_USBPHY);
```
Not part of the PR but this looks wrong. Shouldn't this be
```suggestion
	LL_AHB2_GRP1_DisableClockStopSleep(LL_AHB2_GRP1_PERIPH_OTG_HS |
						LL_AHB2_GRP1_PERIPH_USBPHY);
```

etienne-lms replied:
good catch!

mathieuchopstm commented on drivers/usb/udc/udc_stm32.c:
```diff
@@ -1143,19 +1143,15 @@ static int priv_clock_enable(void)
-	LL_AHB2_GRP1_EnableClock(LL_AHB2_GRP1_PERIPH_USBPHY);
+	clock_control_on(clk, (clock_control_subsys_t)&phy_pclken[0]);
 	/* Both OTG HS and USBPHY sleep clock MUST be disabled here at the same time */
 	LL_AHB2_GRP1_DisableClockStopSleep(LL_AHB2_GRP1_PERIPH_OTG_HS ||
 						LL_AHB2_GRP1_PERIPH_USBPHY);
```
Ditto:
```suggestion
	LL_AHB2_GRP1_DisableClockStopSleep(LL_AHB2_GRP1_PERIPH_OTG_HS |
						LL_AHB2_GRP1_PERIPH_USBPHY);
```

mathieuchopstm commented on dts/arm/st/f7/stm32f7.dtsi:
```diff
@@ -894,6 +894,7 @@
 	otghs_fs_phy: otghs_fs_phy {
 		compatible = "usb-nop-xceiv";
 		#phy-cells = <0>;
+		clocks = <&rcc STM32_CLOCK(AHB1, 30U)>;
```
```suggestion
		clocks = <&rcc STM32_CLOCK(AHB1, 30)>;
```

mathieuchopstm commented on dts/arm/st/h7/stm32h745.dtsi:
```diff
@@ -125,6 +125,7 @@
 	otghs_fs_phy: otghs_fs_phy {
 		compatible = "usb-nop-xceiv";
 		#phy-cells = <0>;
+		clocks = <&rcc STM32_CLOCK(AHB1, 26U)>;
```
```suggestion
		clocks = <&rcc STM32_CLOCK(AHB1, 26)>;
```

mathieuchopstm commented on drivers/usb/device/usb_dc_stm32.c:
```diff
@@ -389,28 +391,24 @@ static int usb_dc_stm32_clock_enable(void)
-#else
-	LL_AHB1_GRP1_EnableClock(LL_AHB1_GRP1_PERIPH_OTGHSULPI);
-#endif
+	clock_control_on(clk, (clock_control_subsys_t)&phy_pclken[0]);
```
I find the duplication of this call all over the place confusing to read.
Is there anything preventing this `clock_control_on` call from being placed in a single spot, gated behind the same `#if USB_OTG_HS_ULPI_PHY || DT_HAS_COMPAT_STATUS_OKAY(st_stm32u5_otghs_phy) || \
USB_OTG_HS_EMB_PHYC` as the `phy_pclken` definition?

(If this is possible, having a top-level `#define HAS_PHY_CLK 1` gated behind the existing `#if` test with all series/IP, and testing for `HAS_PHY_CLK` everywhere else, could be cleaner)

etienne-lms commented on drivers/usb/udc/udc_stm32.c:
```diff
@@ -1147,28 +1147,25 @@ static int priv_clock_enable(void)
 #if USB_OTG_HS_ULPI_PHY
-#if defined(CONFIG_SOC_SERIES_STM32H7X)
+	clock_control_on(clk, (clock_control_subsys_t)&phy_pclken[0]);
 	LL_AHB1_GRP1_EnableClock(LL_AHB1_GRP1_PERIPH_USB1OTGHSULPI);
```
Remove line 1151 (duplicates the `clock_control_on()` call above)

etienne-lms commented on drivers/usb/udc/udc_stm32.c:
```diff
@@ -1147,28 +1147,25 @@ static int priv_clock_enable(void)
+	clock_control_on(clk, (clock_control_subsys_t)&phy_pclken[0]);
 	/* Both OTG HS and USBPHY sleep clock MUST be disabled here at the same time */
-	LL_AHB2_GRP1_DisableClockStopSleep(LL_AHB2_GRP1_PERIPH_OTG_HS ||
+	LL_AHB2_GRP1_DisableClockStopSleep(LL_AHB2_GRP1_PERIPH_OTG_HS |
```
This change is a fix. It deserves a dedicated commit IMHO.

etienne-lms commented on drivers/usb/device/usb_dc_stm32.c:
```diff
@@ -389,28 +391,24 @@ static int usb_dc_stm32_clock_enable(void)
+	clock_control_on(clk, (clock_control_subsys_t)&phy_pclken[0]);
 	/* Both OTG HS and USBPHY sleep clock MUST be disabled here at the same time */
-	LL_AHB2_GRP1_DisableClockStopSleep(LL_AHB2_GRP1_PERIPH_OTG_HS ||
+	LL_AHB2_GRP1_DisableClockStopSleep(LL_AHB2_GRP1_PERIPH_OTG_HS |
```
Fixup change, deserves a dedicated commit.

etienne-lms reviewed (commented):
The 3 last commits changing DTS/DTSI files should be placed before the 2 commits modifying the drivers implementation.

kbidani committed:
drivers: usb: device_stm32: corrected the logical OR || to |

corrected the logical OR (||) to bitwise OR (|) in the
LL_AHB2_GRP1_DisableClockStopSleep function call.

Signed-off-by: Khaoula Bidani <khaoula.bidani-ext@st.com>

kbidani committed:
drivers: usb: udc_stm32: corrected the logical OR || to |

corrected the logical OR (||) to bitwise OR (|) in the
LL_AHB2_GRP1_DisableClockStopSleep function call.

Signed-off-by: Khaoula Bidani <khaoula.bidani-ext@st.com>

etienne-lms commented on drivers/usb/device/usb_dc_stm32.c:
```diff
@@ -101,6 +101,16 @@ static const struct gpio_dt_spec ulpi_reset =
+static const struct stm32_pclken phy_pclken[] = STM32_DT_CLOCKS(DT_INST_PHANDLE(0, phys));
+#endif
+
+#if USB_OTG_HS_ULPI_PHY || DT_HAS_COMPAT_STATUS_OKAY(st_stm32_otghs) || USB_OTG_HS_EMB_PHYC
```
s/st_stm32_otghs/st_stm32u5_otghs_phy/?
If so, can define  `HAS_PHY_CLK` in the above `#ifdef/#endif` close.

etienne-lms commented on drivers/usb/udc/udc_stm32.c:
```diff
@@ -1119,30 +1123,23 @@ static int priv_clock_enable(void)
-#else
-	LL_AHB1_GRP1_EnableClock(LL_AHB1_GRP1_PERIPH_OTGHSULPI);
+#if HAS_PHY_CLK
+	clock_control_on(clk, (clock_control_subsys_t)&phy_pclken[0]);
```
`phy_pclken` definition seems missing in this commit.

etienne-lms commented on drivers/usb/udc/udc_stm32.c:
```diff
@@ -968,6 +968,10 @@ static const struct udc_api udc_stm32_api = {
 #define USB_OTG_HS_ULPI_PHY (DT_HAS_COMPAT_STATUS_OKAY(usb_ulpi_phy) && \
 			    DT_HAS_COMPAT_STATUS_OKAY(st_stm32_otghs))
 
+			    #if USB_OTG_HS_ULPI_PHY || DT_HAS_COMPAT_STATUS_OKAY(st_stm32_otghs) || USB_OTG_HS_EMB_PHYC
```
`#if USB_OTG_HS_ULPI_PHY || USB_OTG_HS_EMB_PHYC` should be enough.

etienne-lms commented on drivers/usb/device/usb_dc_stm32.c:
```diff
@@ -102,13 +102,9 @@ static const struct gpio_dt_spec ulpi_reset =
-#if USB_OTG_HS_ULPI_PHY || DT_HAS_COMPAT_STATUS_OKAY(st_stm32_otghs) || USB_OTG_HS_EMB_PHYC
+#if USB_OTG_HS_ULPI_PHY || USB_OTG_HS_EMB_PHYC
 #define HAS_PHY_CLK 1
+static const struct stm32_pclken phy_pclken[] = STM32_DT_CLOCKS(DT_INST_PHANDLE(0, phys));
```
These 2 changes (line 105 and 107) belong to another commit in the series.

etienne-lms commented on drivers/usb/device/usb_dc_stm32.c:
```diff
@@ -400,31 +404,26 @@ static int usb_dc_stm32_clock_enable(void)
-	LL_AHB1_GRP1_EnableClock(LL_AHB1_GRP1_PERIPH_USB1OTGHSULPI);
-#else
-	LL_AHB1_GRP1_EnableClock(LL_AHB1_GRP1_PERIPH_OTGHSULPI);
+#if HAS_PHY_CLK
```
Looking again, I'm not sure `HAS_PHY_CLK` can be used here.

In previous implementaion, line 416 is embedded when both compats `st_stm32u5_otghs_phy` AND `st_stm32_otghs` are enabled, that is when `USB_OTG_HS_EMB_PHY` is true.
So I think here it should be:
```c
#if USB_OTG_HS_ULPI_PHY || USB_OTG_HS_EMB_PHYC || USB_OTG_HS_EMB_PHY
```
At line 105 (new implementation), you still need to define `phy_pclken` upon
```c
#if USB_OTG_HS_ULPI_PHY || USB_OTG_HS_EMB_PHYC || DT_HAS_COMPAT_STATUS_OKAY(st_stm32u5_otghs_phy)
```
since it's required by `usb_dc_stm32u5_phy_clock_select()` and `usb_dc_stm32u5_phy_clock_enable()`.

Could you cross-check that?

kbidani replied:
yes it's required for u5.

etienne-lms commented on drivers/usb/udc/udc_stm32.c:
```diff
@@ -968,6 +968,11 @@ static const struct udc_api udc_stm32_api = {
 #define USB_OTG_HS_ULPI_PHY (DT_HAS_COMPAT_STATUS_OKAY(usb_ulpi_phy) && \
 			    DT_HAS_COMPAT_STATUS_OKAY(st_stm32_otghs))
 
+#if USB_OTG_HS_ULPI_PHY || USB_OTG_HS_EMB_PHYC
```
` || USB_OTG_HS_EMB_PHY`

etienne-lms replied:
And `USB_OTG_HS_EMB_PHYC` should be defined:
```c
#define USB_OTG_HS_EMB_PHYC (DT_HAS_COMPAT_STATUS_OKAY(st_stm32_usbphyc) && \
			     DT_HAS_COMPAT_STATUS_OKAY(st_stm32_otghs))
```

etienne-lms replied:
Sorry my previous comment was inaccurate and confusing.

For consistency with **device/usb_dc_stm32.c** implementation, I think we should add commit to rename macro `USB_OTG_HS_EMB_PHY` to `USB_OTG_HS_EMB_PHYC` in this **udc/udc_stm32.c** and then introduce `USB_OTG_HS_EMB_PHY` related to STM32U5 OTGHS PHY, as done in **usb_dc_stm32.c**:
```patch
 #define USB_OTG_HS_EMB_PHYC (DT_HAS_COMPAT_STATUS_OKAY(st_stm32_usbphyc) && \
 			      DT_HAS_COMPAT_STATUS_OKAY(st_stm32_otghs))
 
+#define USB_OTG_HS_EMB_PHY (DT_HAS_COMPAT_STATUS_OKAY(st_stm32u5_otghs_phy) && \
+			     DT_HAS_COMPAT_STATUS_OKAY(st_stm32_otghs))
+
 #define USB_OTG_HS_ULPI_PHY (DT_HAS_COMPAT_STATUS_OKAY(usb_ulpi_phy) && \
 			      DT_HAS_COMPAT_STATUS_OKAY(st_stm32_otghs))
```
Based on that, your can factorize PHY clock management upon `USB_OTG_HS_ULPI_PHY || USB_OTG_HS_EMB_PHYC || USB_OTG_HS_EMB_PHY`.

Renaming `USB_OTG_HS_EMB_PHY` to `USB_OTG_HS_EMB_PHYC` could be controversial. @erwango, what do you think.

etienne-lms replied:
Ok, I understand renaming is not the preferred way.
However I think current proposal won't match: `USB_OTG_HS_EMB_PHYC` is not defined (`..._PHY` not `..._PHYC`) and U5 case is not considered (at line 1135 in previous implementation). 

That would be something like:
```patch
 #define USB_OTG_HS_EMB_PHY (DT_HAS_COMPAT_STATUS_OKAY(st_stm32_usbphyc) && \
 			    DT_HAS_COMPAT_STATUS_OKAY(st_stm32_otghs))				    
 
 #define USB_OTG_HS_ULPI_PHY (DT_HAS_COMPAT_STATUS_OKAY(usb_ulpi_phy) && \
 			    DT_HAS_COMPAT_STATUS_OKAY(st_stm32_otghs))				    
 
+#define USB_OTG_HS_EMB_U5_PHY (DT_HAS_COMPAT_STATUS_OKAY(st_stm32u5_otghs_phy) && \
+				DT_HAS_COMPAT_STATUS_OKAY(st_stm32_otghs))
+
-#if USB_OTG_HS_ULPI_PHY || USB_OTG_HS_EMB_PHYC
+#if USB_OTG_HS_ULPI_PHY || USB_OTG_HS_EMB_PHY || USB_OTG_HS_EMB_U5_PHY
 #define HAS_PHY_CLK
```

That said, you proposal implemented in the next commit of these series could apply here, no?
```patch
-#if USB_OTG_HS_ULPI_PHY || USB_OTG_HS_EMB_PHYC
+#if DT_NODE_HAS_PROP(DT_PHANDLE(DT_INST(0, DT_DRV_COMPAT), phys), clocks)
#define HAS_PHY_CLK

kbidani replied:
replaced with: #if DT_NODE_HAS_PROP(DT_PHANDLE(DT_INST(0, DT_DRV_COMPAT), phys), clocks)

etienne-lms commented on drivers/usb/udc/udc_stm32.c:
```diff
@@ -1119,30 +1124,26 @@ static int priv_clock_enable(void)
-	LL_AHB1_GRP1_EnableClock(LL_AHB1_GRP1_PERIPH_USB1OTGHSULPI);
-#else
-	LL_AHB1_GRP1_EnableClock(LL_AHB1_GRP1_PERIPH_OTGHSULPI);
+#if HAS_PHY_CLK
```
`#ifdef`

etienne-lms commented on drivers/usb/udc/udc_stm32.c:
```diff
@@ -1119,30 +1124,26 @@ static int priv_clock_enable(void)
+	}
 #endif
-#elif DT_HAS_COMPAT_STATUS_OKAY(st_stm32_otghs) /* USB_OTG_HS_ULPI_PHY */
+#if DT_HAS_COMPAT_STATUS_OKAY(st_stm32_otghs) /* USB_OTG_HS_ULPI_PHY */
```
By the way, I think an empty line above would be welcome.

etienne-lms commented on drivers/usb/device/usb_dc_stm32.c:
```diff
@@ -400,31 +404,26 @@ static int usb_dc_stm32_clock_enable(void)
+	}
 #endif
-#elif DT_HAS_COMPAT_STATUS_OKAY(st_stm32_otghs) /* USB_OTG_HS_ULPI_PHY */
+#if DT_HAS_COMPAT_STATUS_OKAY(st_stm32_otghs)
```
Here also I think an empty line above would be welcome.

erwango commented on dts/arm/st/h7/stm32h747Xi_m7.dtsi:
```diff
@@ -25,4 +25,10 @@
 			interrupts = <125 0>;
 		};
 	};
+
+	otghs_ulpi_phy: otghs_ulpis_phy {
+		compatible = "usb-ulpi-phy";
+		#phy-cells = <0>;
+		clocks = <&rcc STM32_CLOCK(AHB1, 26)>;
+	};
```
To be moved to stm32h747.dtsi, or to be checked if it wouldn't be obtained by the already available inlcusion of stm32h745.dtsi.

kbidani replied:
Done in stm32h745 dtsi file.

kbidani replied:
moved to stm32h747 dtsi file.

erwango commented on boards/arduino/portenta_h7/arduino_portenta_h7-common.dtsi:
```diff
@@ -24,6 +24,7 @@
 		compatible = "usb-ulpi-phy";
 		reset-gpios = < &gpioj 4 GPIO_ACTIVE_LOW >;
 		#phy-cells = <0>;
+		clocks = <&rcc STM32_CLOCK(AHB1, 26)>;
```
Part of phy descriptions that are not board specifics should be moved to soc files

erwango commented on boards/arduino/nicla_vision/arduino_nicla_vision_stm32h747xx_m7.dts:
```diff
@@ -32,6 +32,7 @@
 
 	otghs_ulpi_phy: otghs_ulpis_phy {
 		compatible = "usb-ulpi-phy";
+		clocks = <&rcc STM32_CLOCK(AHB1, 26)>;
```
ditto

kbidani commented on dts/arm/st/h7/stm32h757Xi_m7.dtsi:
```diff
@@ -31,4 +31,8 @@
 			interrupts = <125 0>;
 		};
 	};
+
+	otghs_ulpi_phy: otghs_ulpis_phy {
+		clocks = <&rcc STM32_CLOCK(AHB1, 26)>;
+	};
```
Done in stm32h745 dtsi file.

kbidani commented on dts/arm/st/f7/stm32f7.dtsi:
```diff
@@ -894,6 +894,7 @@
 		compatible = "usb-nop-xceiv";
 		#phy-cells = <0>;
+		clocks = <&rcc STM32_CLOCK(AHB1, 30)>;
 	};
```
for stm32f723 and stm32f769 status = "okay"
